atox/src/main/kotlin/ActionReceiver.kt

```diff
@@ -71,52 +79,52 @@ class ActionReceiver : BroadcastReceiver() {
-                            Toast.makeText(
-                                context,
-                                context.getString(R.string.error_simultaneous_calls),
-                                Toast.LENGTH_LONG
-                            ).show()
-                            notificationHelper.showPendingCallNotification(UserStatus.Busy, contact)
-                        }
-                        return@launch
-                    }
-
-                    try {
-                        callManager.startCall(pk)
-                        notificationHelper.showOngoingCallNotification(contact)
-                    } catch (e: ToxavAnswerException) {
-                        Log.e(TAG, e.toString())
-                        return@launch
-                    }
-
-                    val isSendingAudio =
-                        context.hasPermission(Manifest.permission.RECORD_AUDIO) && callManager.startSendingAudio()
-                    if (!isSendingAudio) {
-                        withContext(Dispatchers.Main) {
-                            Toast.makeText(context, R.string.call_mic_permission_needed, Toast.LENGTH_LONG).show()
-                        }
-                    }
-                }
-                "reject", "end call" -> {
-                    callManager.endCall(pk)
-                    notificationHelper.dismissCallNotification(pk)
-                }
-                "ignore" -> callManager.removePendingCall(pk)
+        val pk = intent.getStringExtra(KEY_CONTACT_PK)?.let { PublicKey(it) } ?: return
+        when (intent.getSerializableExtra(KEY_ACTION) as Action?) {
+            Action.CallAccept -> acceptCall(context, pk)
+            Action.CallEnd, Action.CallReject -> {
+                callManager.endCall(pk)
+                notificationHelper.dismissCallNotification(pk)
+            }
+            Action.CallIgnore -> callManager.removePendingCall(pk)
+            Action.MarkAsRead -> scope.launch {
+                contactRepository.setHasUnreadMessages(pk.string(), false)
+                notificationHelper.dismissNotifications(pk)
+            }
+            null -> Log.e(TAG, "Missing action in intent $intent")
+        }
+    }
+
+    private fun acceptCall(context: Context, pk: PublicKey) = scope.launch {
+        val contact = contactManager.get(pk).firstOrNull().let {
+            if (it != null) {
+                it
+            } else {
+                Log.e(TAG, "Unable to get contact ${pk.fingerprint()} for call notification")
+                Contact(publicKey = pk.string(), name = pk.fingerprint())
+            }
+        }
+
+        if (callManager.inCall.value is CallState.InCall) {
+            withContext(Dispatchers.Main) {
+                Toast.makeText(context, context.getString(R.string.error_simultaneous_calls), Toast.LENGTH_LONG).show()
+                notificationHelper.showPendingCallNotification(UserStatus.Busy, contact)
```

Comment:
I don't understand this. Does it need a seperate function like `removePendingCallNotification`? Does it suppose to remove the notification?
I assume that it meant to at least block accepting another call. If it does then it still doesn't apply when you initiate a call yourself.

Reply:
This is just code I moved, but the pending call notification is the notification displayed between a call being registered in aTox and it being accepted. Right now the name's a bit odd since it only handles incoming calls, but I'd like to add a notification for pending outgoing calls as well.

All call notifications for a single contact have the same ID as they all are mutually exclusive, so they're all dismissed with the same function. This is in part because if something goes wrong and a notification isn't dismissed, it'll remain until Android's rebooted, so having a easy way to dismiss now-uninteresting calls felt important.

Reply:
OK. But still to note, what this specific line is about? I will also open an issue about the bug I described here.

BTW, you can just dismiss annoying notifications by clicking `Force Stop` in the application's settings.

Reply:
If you're already in call and try to accept a new one, it will tell you off and display the incoming call notification again (but without sound!) as it'll get auto-dismissed when an action is triggered on it.

re: bug, aTox should already be blocking you from starting calls in the chat view if you're already in one. What's the way to start one while already in a call?

It used to be that you couldn't ever dismiss notifications with `setOngoing(true)` on them unless the app did it, but maybe that's changed then?

Reply:
Got it. Forgot about the notification being dimissed when clicking a notification.

re: `re: bug` discussion will continue in the issue.

Yes now it's being dimissed if you force stop it.